File: PhaseJumpPro/Assets/phasejumppro/Interface/UISystem+Drag.cs
using System;
using System.Collections.Generic;
using PJ;
using UnityEngine;

/*
RATING: 5 stars
Tested and works
CODE REVIEW: 12/25/22
*/
namespace PJ
{
    public partial class UISystem
    {
        protected enum DragState
        {
            Default, Drag, LockDragMouseDown, LockDragMouseUp
        }

        /// <summary>
        /// If true, a click will start the drag, and a click will end it
        /// </summary>
        public bool lockDrag = false;

        protected DragModel dragModel;
        protected WeakReference<GameObject> dropTargetOverObject;
        protected DragState dragState = DragState.Default;

        public DragItems DraggedItems
        {
            get
            {
                if (null != dragModel)
                {
                    var list = new List<SomeDragHandler> { dragModel.dragHandler };
                    return new DragItems(list);
                }
                else
                {
                    return new DragItems();
                }
            }
        }

        protected virtual void OnDragUpdate()
        {
            if (!mouseDevice.IsAvailable()) { return; }
            var didMouseButtonUp = mouseDevice.mouse.leftButton.wasReleasedThisFrame;
            var didMouseButtonDown = mouseDevice.mouse.leftButton.wasPressedThisFrame;
            var isMouseButtonDown = mouseDevice.mouse.leftButton.isPressed;

            //if (didMouseButtonDown)
            //{
            //    Debug.Log("isMouseButtonDown: " + isMouseButtonDown.ToString() + "didMouseButtonDown: " + didMouseButtonDown.ToString());
            //}

            if (lockDrag)
            {
                switch (dragState)
                {
                    case DragState.LockDragMouseDown:
                        if (didMouseButtonUp)
                        {
                            dragState = DragState.LockDragMouseUp;
                        }
                        break;
                    case DragState.LockDragMouseUp:
                        if (didMouseButtonDown)
                        {
                            OnDragEnd();
                            return;
                        }
                        break;
                    default:
                        break;
                }
            }

            if (isMouseButtonDown || dragState == DragState.LockDragMouseUp)
            {
                if (null == dragModel) { return; }

                var dragged = dragModel.dragHandler;
                if (null == dragged) { return; }

                CheckDropTargets();

                var inputWorldPosition = mouseDevice.WorldPosition;
                if (null == inputWorldPosition) { return; }

                dragged.OnDragUpdate(new WorldPosition(inputWorldPosition));
            }
            else
            {
                OnDragEnd();
            }
        }

        protected virtual void CheckDropTargets()
        {
            //Debug.Log("Drop: Position: " + mouseDevice.position.ToString());
            var worldPosition = mouseDevice.WorldPosition;
            if (null == worldPosition) { return; }

            var raycastHits = Physics2D.RaycastAll(worldPosition, Vector2.zero);
            DropTarget hitBehavior = null;

            //Debug.Log("Drop: RaycastHits: " + raycastHits.ToString());

            var dragItems = DraggedItems;

            foreach (RaycastHit2D raycastHit in raycastHits)
            {
                if (raycastHit.collider.gameObject == dragModel.dragHandler.gameObject)
                {
                    continue;
                }

                if (raycastHit.collider != null)
                {
                    hitBehavior = raycastHit.collider.gameObject.GetComponent<DropTarget>();
                }

                if (null != hitBehavior && hitBehavior.CanAcceptDrag(dragItems))
                {
                    break;
                }
            }

            bool isAlreadyIn = false;
            if (null != dropTargetOverObject && dropTargetOverObject.TryGetTarget(out GameObject dropTargetTarget))
            {
                var activeDropTarget = dropTargetTarget.GetComponent<DropTarget>();
                if (activeDropTarget != hitBehavior)
                {
                    activeDropTarget.OnDragLeave();
                }
                else
                {
                    isAlreadyIn = true;
                }
            }

            if (null == hitBehavior)
            {
                dropTargetOverObject = null;
                return;
            }

            if (!isAlreadyIn)
            {
                dropTargetOverObject = new WeakReference<GameObject>(hitBehavior.gameObject);
                hitBehavior.OnDragEnter(dragItems);
            }
        }

        public virtual void StartDrag(DragModel dragModel)
        {
            // A mouse button down will start a drag but we don't want that if the drag is locked
            switch (dragState)
            {
                case DragState.LockDragMouseUp:
                    //Debug.Log("Drag is locked. Cancel the current drag first.");
                    return;
            }

            OnDragEnd();
            this.dragModel = dragModel;

            dragState = lockDrag ? DragState.LockDragMouseDown : DragState.Drag;
        }

        public virtual void OnDragEnd()
        {
            if (null != dropTargetOverObject && dropTargetOverObject.TryGetTarget(out GameObject dropTargetTarget))
            {
                var dropTarget = dropTargetTarget.GetComponent<DropTarget>();
                dropTarget.OnDragLeave();

                if (dropTarget.CanAcceptDrag(DraggedItems))
                {
                    dropTarget.OnAcceptDrag(DraggedItems);
                }
            }
            dropTargetOverObject = null;

            if (null == dragModel) { return; }
            dragModel.dragHandler.OnDragEnd();
            dragModel = null;
            dragState = DragState.Default;
        }
    }
}

File: PhaseJump/PhaseJump/Classes/UI/System/UISystem+Drag.cpp
#include "SomeCamera.h"
#include "SomeDragHandler.h"
#include "SomeDropTarget.h"
#include "UIWorldSystem.h"
#include "World.h"

using namespace std;
using namespace PJ;

// TODO: move to mouse motion event?
void UIWorldSystem::OnDragUpdate() {
    if (!IsDragging()) {
        return;
    }
    if (nullptr == mouseDevice || !mouseDevice->IsAvailable()) {
        return;
    }

    auto isMouseButtonDown = mouseDevice->IsButtonDown(PointerInputButton::Left);

    if (isMouseButtonDown || dragState == DragState::LockDragMouseUp) {
        if (nullptr == dragModel) {
            return;
        }

        auto dragged = dragModel->DragHandler();
        GUARD(dragged)

        auto inputScreenPosition = mouseDevice->Position() * world->uiScale;
        auto inputWorldPosition =
            PJ::WorldPosition(Camera()->ScreenToWorld(Vector3(inputScreenPosition)));
        dragged->OnDragUpdate(WorldPosition(inputWorldPosition));
    } else {
        OnDragEnd();
    }
}

void UIWorldSystem::CheckDropTargets(ScreenPosition screenPos) {
    GUARD(dragModel)
    // PJ::Log("Drop: Position: " + mouseDevice->position.ToString());

    auto hits = TestScreenHit(screenPos);
    auto hit = hits.size() > 0 ? &hits[0] : nullptr;
    WorldNode* hitNode = hit ? &hit->node : nullptr;
    auto activeDropTarget = this->activeDropTarget.lock();

    if (activeDropTarget && hitNode != activeDropTarget.get()) {
        DragExitUIEvent event;
        DispatchEvent(*activeDropTarget, SignalId::DragExit, event);
    }

    if (hitNode) {
        activeDropTarget = SCAST<WorldNode>(hitNode->shared_from_this());
        this->activeDropTarget = activeDropTarget;

        DragEnterUIEvent event(*dragModel);
        DispatchEvent(*activeDropTarget, SignalId::DragEnter, event);
    } else {
        activeDropTarget.reset();
    }
}

void UIWorldSystem::StartDrag(SP<DragModel> dragModel) {
    // Don't let hover UI interfere with drag
    SetActiveHover(nullptr);

    // A mouse button down will start a drag but we don't want that if the drag
    // is locked
    switch (dragState) {
    case DragState::LockDragMouseUp:
        // Debug.Log("Drag is locked. Cancel the current drag first.");
        return;
    default:
        break;
    }

    OnDragEnd();
    this->dragModel = dragModel;

    dragState = lockDrag ? DragState::LockDragMouseDown : DragState::Drag;
}

void UIWorldSystem::OnDragEnd() {
    GUARD(dragModel && dragModel->DragHandler())

    // Let the drag handler finish before checking drop targets
    dragModel->DragHandler()->OnDragEnd();

    auto activeDropTarget = this->activeDropTarget.lock();
    if (activeDropTarget) {
        auto dropTarget = activeDropTarget->TypeComponent<SomeDropTarget>();
        if (dropTarget) {
            dropTarget->OnDrop(*dragModel);
            dropTarget->OnDragExit();
        }
        this->activeDropTarget.reset();
    }

    dragModel.reset();
    dragState = DragState::Default;
}
